app/src/main/kotlin/compiler/lowlevel/dataflow/Liveness.kt

```diff
@@ -20,52 +16,52 @@ object Liveness {
             .filter { it.key in subset }
             .associate { it.key to it.value.filter { vertex -> vertex in subset }.toSet() }.toMap()
 
-    object LivenessDataFlowAnalyzer : DataFlowAnalyzer<ReferenceSet<Register>>() {
+    object LivenessDataFlowAnalyzer : DataFlowAnalyzer<Set<Register>>() {
         override val backward = true
 
-        override val entryPointValue: ReferenceSet<Register> = referenceHashSetOf()
+        override val entryPointValue: Set<Register> = setOf()
 
-        override val latticeMaxElement: ReferenceSet<Register> = referenceHashSetOf()
+        override val latticeMaxElement: Set<Register> = setOf()
 
-        override fun latticeMeetFunction(elements: Collection<ReferenceSet<Register>>): ReferenceSet<Register> {
-            return combineReferenceSets(elements.toList())
+        override fun latticeMeetFunction(elements: Collection<Set<Register>>): Set<Register> {
+            return elements.fold(emptySet(), Set<Register>::plus)
         }
 
-        override fun transferFunction(instruction: Instruction, inValue: ReferenceSet<Register>): ReferenceSet<Register> {
+        override fun transferFunction(instruction: Instruction, inValue: Set<Register>): Set<Register> {
             val gen = instruction.regsUsed
             val kill = instruction.regsDefined
-            return combineReferenceSets(referenceHashSetOf(gen.toList()), referenceHashSetOf(inValue.filter { it !in kill }))
+            return gen.toSet() + inValue.filter { it !in kill }.toSet()
         }
     }
 
     fun computeLiveness(linearProgram: List<Asmable>): LivenessGraphs {
         val instructionList = linearProgram.filterIsInstance<Instruction>()
 
         // find all registers, prepare empty graphs
-        val allRegisters = referenceHashSetOf<Register>()
+        val allRegisters = mutableSetOf<Register>()
         instructionList.forEach {
             allRegisters.addAll(it.regsDefined)
             allRegisters.addAll(it.regsUsed)
         }
 
-        val interferenceGraph = referenceHashMapOf(allRegisters.map { it to referenceHashSetOf<Register>() })
-        val copyGraph = referenceHashMapOf(allRegisters.map { it to referenceHashSetOf<Register>() })
+        val interferenceGraph = mutableMapOf(*allRegisters.map { it to mutableSetOf<Register>() }.toTypedArray())
```

Comment:
Maybe we should have another map builder that will take a list explicitly? The conversion to an array followed by spreading is not readable.

Reply:
This is only because Kotlin has no `toMutableMap` for some reason.